rust-symcrypt/src/hash.rs

```diff
@@ -74,47 +115,74 @@ pub fn sha256(data: &[u8]) -> [u8; SHA256_RESULT_SIZE] {
     result
 }
 
+/// Sha384State needs to have a heap allocated inner state that is Pin<Box<T>> since the memory address of Self is moved around when implementing
+/// HashState Result field.
+///
+/// SymCrypt expects the address for its structs to stay static through the structs lifetime to guarantee that structs are not memcpy'd as
+/// doing so would lead to use-after-free and inconsistent states.
+struct Sha384InnerState(symcrypt_sys::SYMCRYPT_SHA384_STATE);
+
 pub struct Sha384State {
-    state: symcrypt_sys::SYMCRYPT_SHA384_STATE,
+    state: Pin<Box<Sha384InnerState>>,
 }
 
+/// Creates a new instance of Sha384State, this must be called before other HashState functions can be called
 impl Sha384State {
     pub fn new() -> Self {
         let mut instance = Sha384State {
-            state: symcrypt_sys::SYMCRYPT_SHA384_STATE::default(),
+            state: Box::pin(Sha384InnerState(
+                symcrypt_sys::SYMCRYPT_SHA384_STATE::default(),
+            )),
```

Comment:
The `#[derive(Default)]` trick would work here too.

Reply:
See other comment on `#[derive(Default)]`